### PR TITLE
Configure Doxygen

### PR DESCRIPTION
**Pull request by Thar0**
Provides a Doxygen configuration in `Doxyfile`.

Predefines `NON_MATCHING` when preprocessing source files to capture non-matching function declarations in the generated documentation.
Filters `#pragma GLOBAL_ASM("path/to/func_name.s")` directives into `? func_name(?)` so that these functions will appear in those files in the generated documentation.

To use it ensure doxygen is installed ( `sudo apt-get install doxygen` ) and run `doxygen Doxyfile` while in the root directory of this project to generate the code documentation.

### CODE REVIEW

File: docs/guides/Documenting.md [hunk 1]
```diff
@@ -0,0 +1,76 @@
...
+ */
+```
+
+## Other
+
+### Documenting Bugs:
+Bugs should be documented on the line above where the bug begins.
+```c
+//! @BUG description
+```
```
Comment (fig02) [inline docs/guides/Documenting.md:66]: currently our code has "bug" in lowercase. does it need to be in uppercase for doxygen?

Comment (Thar0) [reply docs/guides/Documenting.md:66]: lowercase works too if that's preferred

Comment (fig02) [reply docs/guides/Documenting.md:66]: yeah I think its probably what we want to go with since its what we have already, but i dont really mind either way

**Comment by mzxrules** [comment]
could we use /// comments instead

**Pull request merged**